tracing/span.go

```diff
@@ -209,192 +163,492 @@ func (s *Span) ToZipkinSpan() ZipkinSpan {
+	// registered to the parent Span.
+	//
+	// component is an optional, local component name and may be empty string.
+	//
+	// Timestamps, counters, and tags won't be inherited.
+	// Parent id will be inherited from the span id,
+	// and span id will be randomly generated.
+	// Trace id, sampled, and flags will be copied over.
+	CreateLocalChild(name, component string) *LocalSpan
+
+	// CreateLocalChild creates a *LocalSpan with given name and component that
+	// is a child of the Span, starts it, and calls all OnCreateChildHooks
+	// registered to the parent Span as well as a Context that can be used by
+	// the client you are creating the span for.
+	//
+	// component is an optional, local component name and may be empty string.
+	//
+	// Timestamps, counters, and tags won't be inherited.
+	// Parent id will be inherited from the span id,
+	// and span id will be randomly generated.
+	// Trace id, sampled, and flags will be copied over.
+	CreateLocalChildForContext(ctx context.Context, name, component string) (context.Context, *LocalSpan)
+}
+
+func startSpan(s Span) {
+	for _, hook := range s.Hooks() {
+		if err := hook.OnStart(s); err != nil {
+			s.LogError("OnStart hook error: ", err)
 		}
 	}
-	if s.isDebugSet() {
-		s.SetTag(ZipkinBinaryAnnotationKeyDebug, true)
-	}
-	if s.spanType == SpanTypeLocal && s.component != "" {
-		s.SetTag(ZipkinBinaryAnnotationKeyLocalComponent, s.component)
+}
+
+// CreateServerSpan creates a new Server Span, calls any registered
+// BaseplateHooks, and starts the Span.
+func CreateServerSpan(tracer *Tracer, name string) *ServerSpan {
+	span := &ServerSpan{
+		baseSpan{
+			trace: newTrace(tracer, name),
+		},
 	}
-	for _, hook := range s.hooks {
-		if hookErr := hook.OnEnd(s, err); hookErr != nil && s.tracer.Logger != nil {
-			s.tracer.Logger("OnEnd hook error: " + hookErr.Error())
-		}
+	onCreateServerSpan(span)
+	startSpan(span)
+	return span
+}
+
+// GetServerSpan gets the ServerSpan from the given context.
+func GetServerSpan(ctx context.Context) *ServerSpan {
+	if span, ok := ctx.Value(serverSpanKey).(*ServerSpan); ok {
+		return span
 	}
-	if s.tracer != nil {
-		return s.tracer.Record(ctx, s)
+	return nil
+}
+
+// GetChildSpan gets the child Span from the given context.
+func GetChildSpan(ctx context.Context) Span {
+	if span, ok := ctx.Value(childSpanKey).(Span); ok {
+		return span
 	}
 	return nil
 }
 
-// AddCounter adds delta to a counter annotation.
-func (s *Span) AddCounter(key string, delta float64) {
-	s.counters[key] += delta
+// baseSpan provides the default implementation for all methods on the Span
+// interface and is embedded in the exported Span types.
+type baseSpan struct {
+	trace *trace
+	hooks []SpanHook
+}
+
+// Name returns the name of the Span.
+func (s baseSpan) Name() string {
+	return s.trace.name
+}
+
+// SpanType returns the SpanType of the Span.
+func (s baseSpan) SpanType() SpanType {
+	return SpanTypeUnknown
+}
```

Comment:
since `baseSpan` is only used to be embedded into concrete types for code sharing, there's really no need to implement the functions that will be overridden by every concrete types, which includes `SpanType`, `PreStop` and `Stop`.

Reply:
Unfortunately, we have to provide some sort of implementation of we want to actually call into the `Hooks` in the implementations provided here because then `baseSpan` has to adhere to the `Span` interface.

Reply:
oh I see what's happening here now. I think calling the hooks with `baseSpan` is actually incorrect as the hook implementation cannot cast that back into the actual `ServerSpan`/`ClientSpan`/`LocalSpan`.